Jetpacks

KingBalDoro opened the pull request:
is in a half finished state at the time of uploading.

InfectedGriffon committed:
iron lass take one

InfectedGriffon committed:
my hands hurt

kirderf1 committed:
Remove commented-out fillItemCategory() functions

InfectedGriffon committed:
1.20-ify everything

KingBalDoro committed:
Merge branch 'pr/507' into jetpacks

KingBalDoro committed:
Merge branch '1.20.1' into jetpacks

KingBalDoro committed:
nothing here works

KingBalDoro committed:
Merge branch '1.20.1' into jetpacks

KingBalDoro committed:
pushed the feature forward a bit

fixed a lot of the major bugs and added a few things along with some major and minor tweaks ==
- added new sound effects
- changed the speed by with the jetpack propulsion effects the player
-added the jetpack model
-put the files for the animation and texture in the game, although not currently working

KingBalDoro committed:
pushed the mechanic further

- fixed jetpack position on player
- added inventory tick override
- set the ground work for animation controlling
- refined the joystick item code, now can turn off and on boosting mode

KingBalDoro committed:
jetpack mechanically finalized

all features of the jetpack are now in a working state.

current needs before complete feature completion
- the jetpack needs to animate
- the jetpack textures need to be rendered
-item sprites for both jetpack and flight controllers

KingBalDoro committed:
commit presaging PR

still not in a working state, have realized the explosion timer is having an issue. still needs to be able to change animation states, and still needs to render the texture

KingBalDoro committed:
significant changes

significant changes to animation setup
texture now renders
removed jetpack packet and all relevant networking

kirderf1 commented on src/main/java/com/mraof/minestuck/item/armor/JetPackItem.java:
```diff
@@ -0,0 +1,217 @@
+	
+	public void animationStateChange(ItemStack pStack, Level pLevel, Entity pEntity)
+	{
+		if(!isHoldingControllers((LivingEntity) pEntity) && pLevel instanceof ServerLevel serverLevel)
+		{
+			triggerAnim(pEntity, GeoItem.getOrAssignId(pStack, serverLevel), "jetpack_controller", "idle");
+		}
+		if(isHoldingControllers((LivingEntity) pEntity) && pLevel instanceof ServerLevel serverLevel)
+		{
+			triggerAnim(pEntity, GeoItem.getOrAssignId(pStack, serverLevel), "jetpack_controller", "primed");
+		}
+		if(isHoldingControllers((LivingEntity) pEntity) && isBoostingTagTrue(pStack) && pLevel instanceof ServerLevel serverLevel)
+		{
+			triggerAnim(pEntity, GeoItem.getOrAssignId(pStack, serverLevel), "jetpack_controller", "flight");
+		}
```
As you need the instanceof cast in all three blocks, I suggest moving it back to only do it once. Potentially as an early return, or to before calling this function (changing the level paremeter to use `ServerLevel` as type).

KingBalDoro commented on src/main/java/com/mraof/minestuck/data/MinestuckEnUsLanguageProvider.java:
```diff
@@ -1341,6 +1341,8 @@ protected void addTranslations()
 		addItem(MSItems.GOLD_BOAT, "Golden Boat");
 		addItem(MSItems.FLARP_MANUAL, "Flarp Manual");
 		addItem(MSItems.SASSACRE_TEXT, "Sassacre Text");
+		addItem(MSItems.THRUST_CONTROLLER, "Jet Pack Joysticks");
+		addItem(MSItems.JETPACK, "Jetpack");
```
consider it done

KingBalDoro commented on src/main/java/com/mraof/minestuck/item/MSCreativeTabs.java:
might be from when i merged, do you still want it removed?

Dweblenod replied:
Those items are already in mraof/1.20.1 so removing them from your branch would mean reverting their inclusion. Commit history may need to be fixed at the end of this PR. Since this is also reliant on the iron lass armor PR, it may be worthwhile to wait on such Commit fixes since it will need to be done anyway for this PR to accommodate the Commits of the iron lass branch

KingBalDoro commented on src/main/java/com/mraof/minestuck/item/ThrustControllerItem.java:
```diff
@@ -0,0 +1,50 @@
+		return InteractionResultHolder.success(item);
+	}
+	
+	public boolean isBoostingTagTrue(ItemStack stack)
+	{
+		CompoundTag nbt = stack.getTag();
+		
+		return nbt != null && nbt.contains("is_boosting") && nbt.getBoolean("is_boosting");
+	}
```
how

Dweblenod replied:
`jetpackItemStack.isBoostingTagTrue()`

If the question of "how" is about access, you are freely able to access the method in JetPackItem since its set to public. The method here appears to be an exact copy and redundant to that one

KingBalDoro commented on src/main/java/com/mraof/minestuck/item/armor/JetPackItem.java:
```diff
@@ -0,0 +1,217 @@
+	@Override
+	public @NotNull InteractionResultHolder<ItemStack> use(@NotNull Level level, Player player, @NotNull InteractionHand hand)
+	{
+		ItemStack thrustController = new ItemStack(MSItems.THRUST_CONTROLLER.get());
+		
+		if(player.getInventory().getFreeSlot() <= 2)
+		{
+			player.getInventory().add(thrustController);
+		}
```
no, i'm open to suggestions on a better way to go about this.

Dweblenod replied:
Some options I can think of
- Crafting/alchemizing the thrust controllers
- Giving the jetpack a psuedo-storage nbt value to keep track of stored controllers. Similar to how SBURB computer block entities have a value to keep track of blank discs. When the jetpack is first made it will be able to give the player controllers but if the players lose track of said controllers it wont just automatically create new ones unless theyve been stowed back inside the jetpack

KingBalDoro committed:
review edit, rebase, animation fix

- went through some of the reviews on the github
- rebased the entire PR
- fixed the animation switch error. now the only working part of this entire PR
